### DISCUSSION TITLE
Fooocus Private Log

### DISCUSSION THREAD

**Discussion by lllyasviel** [lllyasviel/Fooocus#160] [Announcements]
Fooocus is committed to protect user privacy and **will never write any hidden data in your generated images**.

The option “write hidden meta data” does not even exist in Fooocus, so that you do not need to constantly check if you have forget to turn it on/off, or constantly using some tools to check it.

**Once you are in Fooocus, you can just trust that everything is absolutely private and clean without any hidden things.**

Nevertheless, we also appreciate community feedbacks with the demand to track history generation.

We propose “Fooocus Private Log”, a balance between user privacy protection and tracking flexibility.

After you generate images, you can have a log in the folder “Fooocus/outputs/XXX”, with all detailed parameters in your user interface.

Fooocus is also committed to eliminate the barrier between different software and we encourage users to directly copy prompts as pure texts when sharing them. We will not implement any “fooocus-only” meta data reading functionality to bind users with Fooocus, and users can always try different software directly by copying and pasting your prompts. Those logs are friendly to all other software.

Below is a screenshot of an example log:

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/462923bd-3b18-4888-9df6-96fa3d53a8a6)

![image](https://github.com/lllyasviel/Fooocus/assets/19834515/1ceae0d0-ca41-457d-8ae8-4b2342e7cf85)

Another advantage is that this log can do "Ctrl+F".

**Comment by theelefant** [comment]
This log is missing the seed. Without that information, the log defeats its purpose.

**Reply by lllyasviel** [reply]
fixed in 1.0.32

**Reply by theelefant** [reply]
Thank you for all the effort you've put into this repository. Much appreciated!

**Comment by shingo1228** [comment]
Thanks for implementing private logs. I agree with Fooocus's philosophy of protecting user privacy and not writing hidden data in the generated images, and the way you have implemented it.

**Comment by shingo1228** [comment]
Would it be possible to record the version of "Fooocus" in this log? This would make it easier to track information in the future when new features are added. Such as the 'Sharpness' feature.

**Comment by sakugetu** [comment]
Great solution. Thanks.

**Comment by KrX3D** [comment]
will there be a easy way to get the prompt etc from the log into Fooocus like copy a section from the log and input it somewhere?

**Comment by mattmarkwick** [comment]
I've created a Tampermonkey script that converts the plaintext log file into a gallery app. It won't solve all the problems talked about here but certainly makes browsing created images and finding settings used much easier!

Find it here -> https://github.com/mattmarkwick/fooocus-gallery

![FooocusGallery1](https://github.com/lllyasviel/Fooocus/assets/52632226/089a623b-53de-4b3d-8542-99e936136c05)

**Reply by david-peter-smith** [reply]
I love the project feature and wish a gallery was part of the main Fooocus project but i dislike the use of browser extensions like you used.. wish it was just part of fooocus :(

**Reply by AkoZ** [reply]
this is to be upviewed ,as it is faster and easier to review images in small thumbs than in a big scrolling images log as today.
So may this template be implemented , just the way it is to view a chosen grid (may 4x4 images) in place of one over one with the same parameters copy it is now ?

**Comment by rtfirst** [comment]
Hi Illya,

first of all: You did a really great job! Finally I can play around with the SDXL even on my small RTX-3070 card with only 8GB VRAM. Thanks for that! 😊

I have a small suggestion for improvement:

The Log.html is a good solution, but I it will be better to write each file a seperate log.html with the same filename as the image.png. So it will be easier to find old generations. I switch often and play around with the prompt and sometimes i would like to reproduce older prompts. So at the moment I first need to find the image in the expolorer what is quick but then I need to search in the .html for the filename and the prompt. And that takes more time.

Best regards from germany,
Roland

**Comment by jclevet** [comment]
Can we imagine a way to copy/paste the text of an image's log file into Fooocus to automatically re-parameterize  Fooocus?
Thank you Illya for all your work.

**Reply by metercai** [reply]
maybe, we will "ReGenerate" some finished images at another version of Fooocus: https://github.com/metercai/SimpleSDXL 
the version just updated to view info of finished images, "ReGenerate" will be in next week, like this:
![SimpleSDXL_viewInfo](https://github.com/lllyasviel/Fooocus/assets/5652458/cd544d63-4395-42a3-8628-6211b70fc248)

**Reply by metercai** [reply]
"ReGenerate"  is ok, can try it: https://github.com/metercai/SimpleSDXL
![SimpleSDXL_en1](https://github.com/lllyasviel/Fooocus/assets/5652458/b6eda2b7-e929-44f5-9d31-b6260e0e31cc)

**Comment by Gol-D-Ace** [comment]
What I would like is the private log to fully show the negative and positive prompt that gets fed into the generation.

When you select styles for example it will mention the style name but not the prompt resulting from them.

**Comment by Biglo2001** [comment]
Would be cool if the log would also keep track of the image prompts used, maybe by file name and weight?

**Reply by toutjavascript** [reply]
Absolutely. Image Prompt generation are not reusable in the log.

**Comment by eddyizm** [comment]
If one of the developers could point me to where the image prompt parameters get stored, I can create a PR to add the info to the private log.

**Comment by patrickchamp** [comment]
I view images via built-in photo apps on both PC and a mobile device. A simple txt file alongside each image would be easier to locate using built-in device file explorer apps. Currently, i need to open an HTML file which isn’t supported on mobile devices (e.g. via OneDrive sync, images don’t appear in log depending on file viewer), use CTRL-F (not available in file previews on all sync apps), and in some cases after a file is moved, identify and navigate to the original folder. 

It would be easier to search for and open a txt file sitting alongside each image using a preferred file viewer rather than a more prescriptive log format. This would also make it easier to copy files along with their metadata when moving them. and hopefully would involve less effort on your part to maintain what i’ll note is a pretty nicely formatted log/gallery file.

**Reply by toutjavascript** [reply]
Hello
I made an extension to easily view the images and the prompts :
https://github.com/toutjavascript/Fooocus-Log-Viewer

**Reply by eddyizm** [reply]
@toutjavascript that's pretty neat. I may give it a try on my fork.

**Comment by mashb1t** [comment]
Just for reference: you can find the meta data PR in https://github.com/lllyasviel/Fooocus/pull/1940 (also compatible with [A1111 SD WebUI](https://github.com/AUTOMATIC1111/stable-diffusion-webui)).
Please be advised that this PR is likely not to be merged as it is against the current philosophy, but you can either apply the changes locally or switch to my supported fork https://github.com/mashb1t/Fooocus, which also includes a few more adjustments.

**Reply by patrickchamp** [reply]
Thanks! Yes, I saw that they said they would never merge it. If it instead was an option to save it as a separate txt file with a matching filename they might consider it.

**Reply by toutjavascript** [reply]
I am trying to generate the images with metadata include with my Fooocus Log Viewer
https://github.com/toutjavascript/Fooocus-Log-Viewer
But it is not easy because I only use JS on browser side (for security reason I want a secure module which cannot do anything with python)

**Reply by eddyizm** [reply]
@toutjavascript I may try to integrate your code in my fork when I get a chance.

**Comment by bblackmoor** [comment]
"Fooocus is committed to protect user privacy and will never write any hidden data in your generated images."
-- https://github.com/lllyasviel/Fooocus/discussions/160

I love Fooocus, and I am grateful for all of the time and work put into it. The prompt information is not "hidden data". It is data about the image. Concealing the prompt is concealing essential information about the image.

Allowing a user to choose to include the image metadata (prompt, model, seed, etc.) of the images they generate is no more an invasion of privacy than is the image itself. (It may be helpful to refer to GDPR for a useful definition of "personal data".)
-- https://gdpr-info.eu/issues/personal-data/

Please allow users to **include the image data in the image files** they generate.

Thank you for your time.

**Reply by mashb1t** [reply]
Please find the metadata feature in https://github.com/lllyasviel/Fooocus/releases/tag/2.2.0 and the latest version 2.2.1. You can activate it using the config or the UI (Developer Debug Mode).

**Reply by bblackmoor** [reply]
And it has a choice between text and ison! Thank you and thank lllyasviel. 🙂

**Reply by toutjavascript** [reply]
It should be great if we can add our signature/website/nickname in the metadatas with a free text zone (customizable in config.json)

**Reply by mashb1t** [reply]
@toutjavascript you can already do so, please use the metadata_created_by config!

**Reply by toutjavascript** [reply]
You are the best!

**Comment by datathehuman** [comment]
He showed us, therefore it's true. A contributor to world peace has made another golden post! 
here are some lucky numbers: 29, 6, 36, 10, 41, 40
1 of 1 iteration, from https://www.gigacalculator.com/calculators/lucky-number-generator.php

**Comment by edik1710** [comment]
The log file or prompt is tracker on net.
My administrators of network can see what i write on prompt or in log file if I dont share with nobody?
